k2/csrc/cuda/debug.cuh

```diff
@@ -0,0 +1,169 @@
+ */
+#ifndef K2_CUDA_CHECK_ERROR
+  #define K2_CUDA_CHECK_ERROR(e) \
+    k2::K2_CUDA_DEBUG((cudaError_t) (e), __FILE__, __LINE__, true)
+#endif
+
+/**
+ * \brief Macro for checking cuda standard runtime api return status.
+ * If api return status is not cudaSuccess, print and exit.
+ * Thus it's actually could be an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * \note:
+ * All runtime api return an error code, but for an asynchronous api,
+ * the error code only reports errors that occur on the host
+ * prior to executing the task, typically related to parameter validation.
+ * The device kernel error would unfortunately be left for next runtime api call.
+ * To check cuda runtime async api, as cuda kernel launches are asynchronous also,
+ * the same macro for kernel could be used.
+ * See `K2_CUDA_KERNEL_SAFE_CALL`.
+ *
+ * \usage:
+ * K2_CUDA_API_SAFE_CALL(cuda_runtime_api())
+ */
+#ifndef K2_CUDA_API_SAFE_CALL
+  #define K2_CUDA_API_SAFE_CALL(c) K2_CUDA_CHECK_ERROR(c)
+#endif
+
+/**
+ * \brief Macro for checking cuda standard runtime api return status.
+ * If api return status is not cudaSuccess, print and exit.
+ * Thus it's actually an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * \note:
+ * Kernel launches do not return any error code, thus checking should after it.
+ * To wait kernel to finish, cudaDeviceSynchronize is called between.
+ * And to avoid pre-launch error disturbing, check error before kernel launch is
+ * also needed (but not necessary if one promise each error is take cared properly)
+ *
+ * \usage:
+ * K2_CUDA_KERNEL_SAFE_CALL(kernel_foo<<<>>>())
+ */
+#ifndef K2_CUDA_KERNEL_SAFE_CALL
+  #define K2_CUDA_KERNEL_SAFE_CALL(c)                 \
+    do {                                              \
+      K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+      K2_CUDA_API_SAFE_CALL(c);                       \
+      K2_CUDA_API_SAFE_CALL(cudaDeviceSynchronize()); \
+      K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+    } while (0)
+#endif
+
+/**
+ * \brief Log macro for printf statements.
+ * `printf` is supported by both host and device.
+ * This Log is for debugging, the error msg is printed to the stderr.
+ *
+ * Refer to:
+ * https://github.com/NVlabs/cub/blob/c3cceac115c072fb63df1836ff46d8c60d9eb304/cub/util_debug.cuh#L109
+ */
+#if !defined(K2_DLOG)
+  #if !(defined(__clang__) && defined(__CUDA__))
+    #if (K2_PTX_ARCH == 0)
+      #define K2_DLOG(format, ...) printf(format,__VA_ARGS__)
+    #elif (K2_PTX_ARCH >= 200)
+      #define K2_DLOG(format, ...)                                            \
+        printf("[block (%d,%d,%d), thread (%d,%d,%d)]: " format, blockIdx.z,  \
+               blockIdx.y, blockIdx.x, threadIdx.z, threadIdx.y, threadIdx.x, \
+               __VA_ARGS__)
+#endif
+  #else
+    // A hack to implement the variadic printf as clang for other c-compilers,
+    // and sielence the warning
+    #pragma clang diagnostic ignored "-Wc++11-extensions"
```

Comment:
What happens if a user does not use clang?

What are the purposes of these two pragmas?

If they are really needed it, please wrap it with pragma push and pop.

Reply:
I made a mistake with the comments. This `pragma` is for clang, and you are right, a push and pop guard is added.